Configuring mvn tool potential issue

I'm having issues configuring a maven tool with configuration as code. The issue seems to be that Configuration-as-Code is confused as to what configurator to use. 

We've got this from the configuration:
![tools](https://user-images.githubusercontent.com/1484793/38492350-e12807b2-3bee-11e8-9531-2cd3d9cffab0.png)

So i started out with just trying to get it to work (Maven installations only require a name)

``` 
tool:
  maven:
     installations:
      - name: "myInstallation"
``` 
Which gives me: 

``` 
jenkins_1  | Caused by: java.lang.IllegalArgumentException: Invalid configuration elements for type class jenkins.mvn.GlobalMavenConfig : installations
jenkins_1  | 	at org.jenkinsci.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:200)
jenkins_1  | 	at org.jenkinsci.plugins.casc.DescriptorConfigurator.configure(DescriptorConfigurator.java:45)
jenkins_1  | 	at org.jenkinsci.plugins.casc.DescriptorConfigurator.configure(DescriptorConfigurator.java:12)
jenkins_1  | 	at org.jenkinsci.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:188)
jenkins_1  | 	at org.jenkinsci.plugins.casc.GlobalConfigurationCategoryConfigurator.configure(GlobalConfigurationCategoryConfigurator.java:45)
jenkins_1  | 	at org.jenkinsci.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:208)
jenkins_1  | 	... 28 more
jenkins_1  | 
``` 
So it seems that Confguration as code tries to configure: https://github.com/jenkinsci/jenkins/blob/08def67a18eee51de9f3f99bc2a792fee1c160e0/core/src/main/java/jenkins/mvn/GlobalMavenConfig.java

While...i think it's supposed to configure: https://github.com/jenkinsci/jenkins/blob/d688c154907d17e75ea31067c8fc2525aa679584/core/src/main/java/hudson/tasks/Maven.java#L666

@ndeloof @oleg-nenashev Are my assumption correct, in that we're configuring the wrong thing via casc?

--- a/plugin/src/main/java/org/jenkinsci/plugins/casc/Configurator.java
+++ b/plugin/src/main/java/org/jenkinsci/plugins/casc/Configurator.java
@@ -290,14 +290,6 @@ public abstract class Configurator<T> implements ExtensionPoint, ElementConfigur
     }
 
     /**
-     * Determine the list of Attribute available for configuration of the managed component.
-     *
-     * @return A set of {@link Attribute}s that describes this object
-     */
-    @Nonnull
-    public abstract Set<Attribute> describe();
-
-    /**
      * Retrieve the html help tip associated to an attribute.
      * FIXME would prefer &lt;st:include page="help-${a.name}.html" class="${c.target}" optional="true"/&gt;
      * @param attribute to get help for

--- a/plugin/src/main/java/org/jenkinsci/plugins/casc/core/MavenConfigurator.java
+++ b/plugin/src/main/java/org/jenkinsci/plugins/casc/core/MavenConfigurator.java
@@ -1,0 +1,57 @@
+package org.jenkinsci.plugins.casc.core;
+
+import hudson.Extension;
+import hudson.model.Descriptor;
+import hudson.tasks.Maven;
+import jenkins.model.Jenkins;
+import jenkins.mvn.GlobalMavenConfig;
+import org.jenkinsci.plugins.casc.Attribute;
+import org.jenkinsci.plugins.casc.BaseConfigurator;
+import org.jenkinsci.plugins.casc.Configurator;
+import org.jenkinsci.plugins.casc.model.CNode;
+import org.jenkinsci.plugins.casc.model.Mapping;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.Set;
+
+/**
+ * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
+ */
+@Extension
+public class MavenConfigurator extends BaseConfigurator<GlobalMavenConfig> {
+
+
+    @Override
+    public Class<GlobalMavenConfig> getTarget() {
+        return GlobalMavenConfig.class;
+    }
+
+    @Override
+    protected GlobalMavenConfig instance(Mapping mapping) {
+        return GlobalMavenConfig.get();
+    }
+
+    @Nonnull
+    @Override
+    public Set<Attribute> describe() {
+        final Set<Attribute> attributes = super.describe();
+        final Descriptor descriptor = Jenkins.getInstance().getDescriptorOrDie(Maven.class);
+        final Configurator<Descriptor> task = Configurator.lookup(descriptor.getClass());
+        if (task == null) return attributes; // ?
+
+        for (Attribute attribute : task.describe()) {
+            attributes.add(new Attribute(attribute.getName(), attribute.getType())
+                .multiple(attribute.isMultiple())
+                .getter(g -> attribute.getValue(descriptor))
+                .setter((g,v) -> attribute.setValue(descriptor,v)));
+        }
+        return attributes;
+    }
+
+    @CheckForNull
+    @Override
+    public CNode describe(GlobalMavenConfig instance) throws Exception {
+        return null;
+    }
+}

--- a/plugin/src/test/java/org/jenkinsci/plugins/casc/core/MavenConfiguratorTest.java
+++ b/plugin/src/test/java/org/jenkinsci/plugins/casc/core/MavenConfiguratorTest.java
@@ -1,0 +1,33 @@
+package org.jenkinsci.plugins.casc.core;
+
+import hudson.tasks.Maven;
+import jenkins.mvn.FilePathSettingsProvider;
+import jenkins.mvn.GlobalMavenConfig;
+import jenkins.mvn.SettingsProvider;
+import org.jenkinsci.plugins.casc.misc.ConfiguredWithCode;
+import org.jenkinsci.plugins.casc.misc.JenkinsConfiguredWithCodeRule;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+
+/**
+ * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
+ */
+public class MavenConfiguratorTest {
+
+    @Rule
+    public JenkinsConfiguredWithCodeRule j = new JenkinsConfiguredWithCodeRule();
+
+    @Test
+    @ConfiguredWithCode("MavenConfiguratorTest.yml")
+    public void should_configure_maven_tools_and_global_config() {
+        final Maven.DescriptorImpl descriptor= (Maven.DescriptorImpl) j.jenkins.getDescriptorOrDie(Maven.class);
+        Assert.assertEquals(1, descriptor.getInstallations().length);
+        Assert.assertEquals("/usr/share/maven", descriptor.getInstallations()[0].getHome());
+
+
+        final SettingsProvider provider = GlobalMavenConfig.get().getSettingsProvider();
+        Assert.assertTrue(provider instanceof FilePathSettingsProvider);
+        Assert.assertEquals("/usr/share/maven-settings.xml", ((FilePathSettingsProvider)provider).getPath());
+    }
+}

--- a/plugin/src/test/resources/org/jenkinsci/plugins/casc/core/MavenConfiguratorTest.yml
+++ b/plugin/src/test/resources/org/jenkinsci/plugins/casc/core/MavenConfiguratorTest.yml
@@ -1,0 +1,8 @@
+tool:
+  maven:
+    settingsProvider:
+      filePath:
+        path: /usr/share/maven-settings.xml
+    installations:
+      - name: maven
+        home: /usr/share/maven

